### PR TITLE
decodeIfPresent

### PR DESCRIPTION
**Pull request by valeriomazzeo**

### CODE REVIEW

File: Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift [hunk 1]
```diff
@@ -0,0 +1,74 @@
+//
+//  KeyedEncodingContainerProtocol+AnyCodableTests.swift
+//  AnyCodableTests
+//
+//  Created by Dale Buckley on 22/03/2018.
+//
```
Comment (valeriomazzeo) [inline Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift]: copyright

File: Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        ("testEncodingEmptyAnyCodable", testEncodingEmptyAnyCodable),
+        ("testEncodingNonEmptyAnyCodable", testEncodingNonEmptyAnyCodable)
+    ]
+
+    func testEncodingEmptyAnyCodable() throws {
+
+        let mock = MockCodable(nil)
+        let data = try JSONEncoder().encode(mock)
+
+        XCTAssertFalse(String(data: data, encoding: .utf8)!.contains("\"any\":"))
```
Comment (valeriomazzeo) [inline Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift:28]: should that be checking `null` as well?

File: Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+//
+
+import XCTest
+@testable import AnyCodable
+
+class KeyedEncodingContainerProtocolAnyCodableTests: XCTestCase {
+    
+    static let allTests = [
+        ("testEncodingEmptyAnyCodable", testEncodingEmptyAnyCodable),
+        ("testEncodingNonEmptyAnyCodable", testEncodingNonEmptyAnyCodable)
```
Comment (valeriomazzeo) [inline Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift]: `testEncodingIfPresentWithNilValue`
`testEncodingIfPresent`

File: Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+        let decoded = try JSONDecoder().decode(MockCodable.self, from: data)
+
+        XCTAssertEqual(decoded.any as! String, "test")
+    }
+}
+
+extension KeyedEncodingContainerProtocolAnyCodableTests {
+
+    struct MockCodable: Codable {
```
Comment (valeriomazzeo) [inline Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift:98]: I wonder whether this works also when `Codable` is automatically synthesised.
Probably worth to test:
```
struct MockCodable: Codable {
  let name: String
}
```

File: Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift (hunk 2 above)
Comment (dlbuckley) [reply Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift:28]: The whole point is that `"any"` won't exist, so no. But I will add a test that takes an `NSNull` so we can test that too.

File: Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift (hunk 4 above)
Comment (dlbuckley) [reply Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift:98]: Well that particular case will just use the `String` overload. If we want to see if auto synth will work then it should have a property of `AnyCodable`.

File: Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift (hunk 2 above)
Comment (valeriomazzeo) [reply Tests/AnyCodableTests/KeyedEncodingContainerProtocol+AnyCodableTests.swift:28]: I meant `"\"any\":null"`

**Review by dlbuckley — APPROVED** [review]

**Comment by valeriomazzeo** [comment]
👍

**Pull request merged**